lib/ansible/modules/network/cloudengine/ce_ospf.py

```diff
@@ -0,0 +1,996 @@
+                                (self.addr, self.get_wildcard_mask()))
+        self.changed = True
+
+    def remove_area(self):
+        """remove ospf area"""
+
+        if not self.is_area_exist():
+            return
+
+        xml_area = CE_NC_XML_BUILD_DELETE_AREA % (self.get_area_ip(), "")
+        xml_str = CE_NC_XML_BUILD_PROCESS % (self.process_id, xml_area)
+        recv_xml = set_nc_config(self.module, xml_str)
+        self.check_response(recv_xml, "DELETE_AREA")
+        self.updates_cmd.append("ospf %s" % self.process_id)
+        self.updates_cmd.append("undo area %s" % self.get_area_ip())
+        self.changed = True
+
+    def remove_nexthop(self):
+        """remove ospf nexthop weight"""
+
+        if not self.is_nexthop_exist():
+            return
+
+        xml_nh = CE_NC_XML_DELETE_NEXTHOP % self.nexthop_addr
+        xml_topo = CE_NC_XML_BUILD_TOPO % xml_nh
+        xml_str = CE_NC_XML_BUILD_PROCESS % (self.process_id, xml_topo)
+        recv_xml = set_nc_config(self.module, xml_str)
+        self.check_response(recv_xml, "DELETE_NEXTHOP_WEIGHT")
+        self.updates_cmd.append("ospf %s" % self.process_id)
+        self.updates_cmd.append("undo nexthop %s" % self.nexthop_addr)
+        self.changed = True
+
+    def is_valid_v4addr(self, addr):
+        """check is ipv4 addr is valid"""
+
+        if addr.find('.') != -1:
+            addr_list = addr.split('.')
+            if len(addr_list) != 4:
+                return False
+            for each_num in addr_list:
+                if not each_num.isdigit():
+                    return False
+                if int(each_num) > 255:
+                    return False
+            return True
+
+        return False
+
+    def convert_ip_to_network(self):
+        """convert ip to subnet address"""
+
+        ip_list = self.addr.split('.')
+        mask_list = self.get_wildcard_mask().split('.')
+
+        for i in range(len(ip_list)):
+            ip_list[i] = str((int(ip_list[i]) & (~int(mask_list[i]))) & 0xff)
+
+        self.addr = '.'.join(ip_list)
+
+    def check_params(self):
+        """Check all input params"""
+
+        # process_id check
+        if not self.process_id.isdigit():
+            self.module.fail_json(msg="Error: process_id is not digit.")
+        if int(self.process_id) < 1 or int(self.process_id) > 4294967295:
+            self.module.fail_json(
+                msg="Error: process_id must be an integer between 1 and 4294967295.")
+
+        if self.area:
+            # area check
+            if self.area.isdigit():
+                if int(self.area) < 0 or int(self.area) > 4294967295:
+                    self.module.fail_json(
+                        msg="Error: area id (Integer) must be between 0 and 4294967295.")
+
+            else:
+                if not self.is_valid_v4addr(self.area):
+                    self.module.fail_json(msg="Error: area id is invalid.")
+
+            # area network check
+            if self.addr:
```

Comment:
Does `mask` require `addr`, and does `addr` require `mask`, if so you may wish to add:
```
required_together
```